base/task.jl

```diff
@@ -306,6 +306,18 @@ function _wait2(t::Task, waiter::Task)
         if !istaskdone(t)
             push!(t.donenotify.waitq, waiter)
             unlock(t.donenotify)
+            # since _wait2 is similar to schedule, we should observe the sticky
+            # bit, even if we aren't calling `schedule` due to this early-return
+            if waiter.sticky && Threads.threadid(waiter) == 0
+                # Issue #41324
+                # t.sticky && tid == 0 is a task that needs to be co-scheduled with
+                # the parent task. If the parent (current_task) is not sticky we must
+                # set it to be sticky.
+                # XXX: Ideally we would be able to unset this
+                current_task().sticky = true
+                tid = Threads.threadid()
+                ccall(:jl_set_task_tid, Cvoid, (Any, Cint), waiter, tid-1)
```

Comment:
@vtjnash and @vchuravy: I think this is wrong. Here, `waiter` is waiting on `t`, not on `current_task()` which can be completely different (e.g. see [bind](https://github.com/JuliaLang/julia/blob/master/base/channels.jl#L256)).

Apparently, we want to pin `t` and `waiter` to the same thread. But I'm not sure how this follows from the original problem: #41324. It's perfectly reasonable for a task in the interactive threadpool to wait on a task in the default threadpool, but this makes that unachievable.

I know it's been a while, but can either of you remember why this was added after #41334?

Reply:
`_wait2` is defined as semantically a call to `schedule` the Task to immediately call `wait` on some condition variable. Hence why it has the same side-effects as calling `schedule`: it should pin `current_task` and `waiter` to one thread, but not `t` (which will have completed before `waiter` runs anyways).

Reply:
That task `t` will have completed before `waiter` runs makes sense. But _this_ `_wait2` as opposed to the one in `condition.jl` has nothing to do with `Condition`s. `t` is a `Task` here, not a `Condition`. The semantics of this `_wait2` seem unrelated to `current_task()`.